### PR TITLE
executor: makes dirty table supports XAPI.

### PR DESCRIPTION
**Pull request by coocood**
Currently, if a transaction has written some data, XAPI can not be used, this PR fixes this issue.

### CODE REVIEW

File: executor/union_scan.go [hunk 1]
```diff
@@ -0,0 +1,243 @@
+package executor
```
Comment (shenli) [inline executor/union_scan.go]: Miss license.
Any test case for this?

Comment (coocood) [inline executor/union_scan.go]: I will add tests for this later, but most of the code in this file is already covered by old tests.

**Comment by ngaut** [comment]
PTAL @hanfei19910905 @zimulala

File: executor/builder.go [hunk 2]
```diff
@@ -226,9 +226,14 @@ func (b *executorBuilder) buildTableScan(v *plan.TableScan) Executor {
...
 		if len(remained) == 0 {
 			e.allFiltersPushed = true
-			return e
 		}
-		return b.buildFilter(e, remained)
+		var ex Executor
+		if txn.IsReadOnly() {
+			ex = e
+		} else {
+			ex = b.buildUnionScanExec(e)
```
Comment (shenli) [inline executor/builder.go]: Not all the statements can be pushed down.

File: executor/union_scan.go [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+}
+
+// DirtyDBKey is the key to *dirtyDB for a context.
+const DirtyDBKey dirtyDBKeyType = 1
+
+func getDirtyDB(ctx context.Context) *dirtyDB {
+	var udb *dirtyDB
+	x := ctx.Value(DirtyDBKey)
+	if x == nil {
+		udb = &dirtyDB{tables: make(map[int64]*dirtyTable)}
```
Comment (shenli) [inline executor/union_scan.go]: How about modify two databases in a transaction?

File: store/tikv/mock-tikv/cop_handler.go [hunk 4]
```diff
@@ -264,7 +264,7 @@ func (h *rpcHandler) getRowsFromRange(ctx *selectContext, ran kv.KeyRange, limit
 			if bytes.Compare(pair.Key, startKey) < 0 {
 				break
 			}
-			seekKey = pair.Key
+			seekKey = []byte(tablecodec.TruncateToRowKeyLen(kv.Key(pair.Key)))
```
Comment (shenli) [inline store/tikv/mock-tikv/cop_handler.go]: Is this a bug? I remember that we have fixed it.

File: executor/builder.go [hunk 5]
```diff
@@ -588,3 +599,25 @@ func (b *executorBuilder) buildExplain(v *plan.Explain) Executor {
 		fields:   v.Fields(),
 	}
 }
+
+func (b *executorBuilder) buildUnionScanExec(src Executor) *UnionScanExec {
```
Comment (shenli) [inline executor/builder.go]: Add some comments for this function.

File: executor/union_scan.go [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/evaluator"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/table"
+	"github.com/pingcap/tidb/util/types"
+)
+
+type dirtyDB struct {
```
Comment (shenli) [inline executor/union_scan.go]: Need comments.

File: executor/union_scan.go [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+	tables map[int64]*dirtyTable
+}
+
+func (udb *dirtyDB) AddRow(tid, handle int64, row []types.Datum) {
```
Comment (shenli) [inline executor/union_scan.go]: Can we make this private?

Comment (shenli) [inline executor/union_scan.go]: So as the following functions.

File: executor/union_scan.go [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+		dt = &dirtyTable{
+			addedRows:   make(map[int64][]types.Datum),
+			deletedRows: make(map[int64]struct{}),
+		}
+		udb.tables[tid] = dt
+	}
+	return dt
+}
+
+type dirtyTable struct {
```
Comment (shenli) [inline executor/union_scan.go]: Can we put this into sessionctx package?

File: executor/union_scan.go [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+		udb = &dirtyDB{tables: make(map[int64]*dirtyTable)}
+		ctx.SetValue(DirtyDBKey, udb)
+	} else {
+		udb = x.(*dirtyDB)
+	}
+	return udb
+}
+
+// UnionScanExec merges the rows from dirty table and the rows from XAPI request.
+type UnionScanExec struct {
```
Comment (shenli) [inline executor/union_scan.go]: This could be private?

File: executor/builder.go (hunk 2 above)
Comment (coocood) [inline executor/builder.go]: It's table scan and client support has been checked.

File: executor/union_scan.go [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+		snapshotRow, err := us.Src.Next()
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		var addedRow *Row
+		if us.cursor < len(us.addedRows) {
+			addedRow = us.addedRows[us.cursor]
+		}
+		if us.dirty.truncated || snapshotRow == nil {
+			if addedRow != nil {
```
Comment (shenli) [inline executor/union_scan.go]: How about TruncateTable-> AddRows->TruncateTable->AddRows?

File: executor/union_scan.go [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+				for i, field := range us.Src.Fields() {
+					field.Expr.SetDatum(addedRow.Data[i])
+				}
+				us.cursor++
+				return addedRow, nil
+			}
+			return nil, nil
+		}
+		if addedRow == nil {
```
Comment (shenli) [inline executor/union_scan.go]: Do not consider removed rows?

File: executor/union_scan.go (hunk 8 above)
Comment (coocood) [inline executor/union_scan.go]: No, because it depends on `executor` package.

Comment (coocood) [inline executor/union_scan.go]: Well, it used to depends on `executor` before, but after a refactor, it doesn't now,
But it's more convenient to be used here.

File: executor/union_scan.go (hunk 3 above)
Comment (coocood) [inline executor/union_scan.go]: Table IDs are unique.

File: executor/union_scan.go [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+		}
+		if len(snapshotRow.RowKeys) != 1 {
+			return nil, ErrRowKeyCount
+		}
+		snapshotHandle := snapshotRow.RowKeys[0].Handle
+		if _, ok := us.dirty.deletedRows[snapshotHandle]; ok {
+			continue
+		}
+		if _, ok := us.dirty.addedRows[snapshotHandle]; ok {
+			// If src handle appears in added rows, it means there is conflict and the transaction will fail to
```
Comment (shenli) [inline executor/union_scan.go]: I can not understand. Is this the scenario that we get rows from snapshot which will be modified by this txn?

File: store/tikv/mock-tikv/cop_handler.go (hunk 4 above)
Comment (coocood) [inline store/tikv/mock-tikv/cop_handler.go]: Yes, it has been fixed in localstore, but didn't in mock tikv.

File: mysql/set.go [hunk 13]
```diff
@@ -57,7 +57,7 @@ func ParseSetName(elems []string, name string) (Set, error) {
 		if _, ok := marked[key]; ok {
 			value |= (1 << uint64(i))
 			delete(marked, key)
-			items = append(items, key)
```
Comment (shenli) [inline mysql/set.go]: Is this a bug? Pleas add a test case.

File: executor/union_scan.go (hunk 9 above)
Comment (coocood) [inline executor/union_scan.go]: Yes, but all the executors are public right now, so we better keep it consistent.

File: executor/union_scan.go (hunk 10 above)
Comment (coocood) [inline executor/union_scan.go]: I'll add test case for it.

File: executor/union_scan.go (hunk 11 above)
Comment (coocood) [inline executor/union_scan.go]: Ah, I should check removed rows.

File: executor/union_scan.go (hunk 12 above)
Comment (coocood) [inline executor/union_scan.go]: We first insert a row, with an assumption that there is no conflict, then we select rows from snapshot, found that the row exists.

File: mysql/set.go (hunk 13 above)
Comment (coocood) [inline mysql/set.go]: It is covered by old test case which inserted a set with uppercase string, then get back lowercase string back.

**Comment by coocood** [comment]
@shenli PTAL

File: executor/builder.go (hunk 2 above)
Comment (shenli) [inline executor/builder.go]: How about aggregate function? I will enable aggregate function xapi soon.

Comment (coocood) [inline executor/builder.go]: If the table is dirty, do not push aggregate functions, table scan can still use XAPI.

File: executor/union_scan.go [hunk 14]
```diff
@@ -0,0 +1,293 @@
...
+			return 0, errors.Trace(err)
+		}
+		if cmp != 0 {
+			return cmp, nil
+		}
+	}
+	aHandle := a.RowKeys[0].Handle
+	bHandle := b.RowKeys[0].Handle
+	var cmp int
+	if aHandle == bHandle {
```
Comment (shenli) [inline executor/union_scan.go]: Will this happen?

File: executor/union_scan.go [hunk 15]
```diff
@@ -0,0 +1,293 @@
...
+	return addedRow
+}
+
+func (us *UnionScanExec) pickRow(a, b *Row) (*Row, error) {
+	addedCmpSrc, err := us.compare(a, b)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	var row *Row
+	// Compare result will never be 0.
```
Comment (shenli) [inline executor/union_scan.go]: Why?

Comment (coocood) [inline executor/union_scan.go]: One of the argument is snapshotRow which is checked in added rows before.

File: executor/union_scan.go [hunk 16]
```diff
@@ -0,0 +1,293 @@
...
+		row := &Row{Data: data, RowKeys: []*RowKeyEntry{rowKeyEntry}}
+		us.addedRows = append(us.addedRows, row)
+	}
+	if us.desc {
+		sort.Sort(sort.Reverse(us))
+	} else {
+		sort.Sort(us)
+	}
+	if us.sortErr != nil {
+		return us.sortErr
```
Comment (shenli) [inline executor/union_scan.go]: return errors.Trace(us.sortErr)

File: executor/union_scan.go (hunk 14 above)
Comment (coocood) [inline executor/union_scan.go]: No if we only use it to compare added row and snapshot row.

**Comment by shenli** [comment]
LGTM
But please add some unit test cases latter.

**Comment by coocood** [comment]
@shenli test cases have been added already.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**